### PR TITLE
MINOR: Introduce OffsetAndEpoch in LeaderEndpoint interface return values

### PR DESCRIPTION
**Pull request by kowshik**
Some of the `LeaderEndpoint` interface methods used a tuple of `(Int, Long)` as return value to represent epoch & offset. In this PR, I have replaced those occcurences with the existing `OffsetAndEpoch` type. This should hopefully improve code readability, and eliminate the chances of developer confusing offset with epoch (or vice versa) at the call site or return site.

**Tests:**
Rely on existing tests

### CODE REVIEW

**Comment by kowshik** [comment]
Hi @junrao and @satishd -- Please could you help review this PR?

File: core/src/main/scala/kafka/server/OffsetAndEpoch.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.server
+
+case class OffsetAndEpoch(offset: Long, leaderEpoch: Int) {
```
Comment (junrao) [inline core/src/main/scala/kafka/server/OffsetAndEpoch.scala]: Could we add this in the server-common module in java?

**Review by junrao — COMMENTED** [review]
@kowshik : Thanks for the PR. Left a comment. Also, the built failed with unused import.

**Comment by kowshik** [comment]
@junrao Thanks for the review! I've addressed the comments, the PR is ready for another pass.

**Review by satishd — APPROVED** [review]
Thanks @kowshik for the updated PR, LGTM.

File: core/src/main/scala/kafka/server/AbstractFetcherThread.scala [hunk 2]
```diff
@@ -704,18 +708,18 @@ abstract class AbstractFetcherThread(name: String,
...
+      val leaderStartOffset = offsetAndEpoch.offset
       warn(s"Reset fetch offset for partition $topicPartition from $replicaEndOffset to current " +
         s"leader's start offset $leaderStartOffset")
       val offsetToFetch =
         if (leaderStartOffset > replicaEndOffset) {
           // Only truncate log when current leader's log start offset (local log start offset if >= 3.4 version incaseof
           // OffsetMovedToTieredStorage error) is greater than follower's log end offset.
           // truncateAndBuild returns offset value from which it needs to start fetching.
-          truncateAndBuild(epoch, leaderStartOffset)
+          truncateAndBuild(offsetAndEpoch.leaderEpoch, leaderStartOffset)
```
Comment (Hangleton) [inline core/src/main/scala/kafka/server/AbstractFetcherThread.scala]: Minor - maybe the argument of `truncateAndBuild` could be the `offsetAndEpoch` itself?

Comment (kowshik) [reply core/src/main/scala/kafka/server/AbstractFetcherThread.scala]: Done in 3da1e63. Good point!

**Comment by kowshik** [comment]
@Hangleton Thanks for the review. I have addressed the comment, the PR is ready for another pass.

**Review by junrao — COMMENTED** [review]
@kowshik : Thanks for the updated PR. Could you rebase?

**Comment by kowshik** [comment]
@junrao Thanks for the review. I have rebased the PR and brought in the latest commits from AK trunk. PR is ready for a pass.

**Review by junrao — APPROVED** [review]
@kowshik : Thanks for rebasing the PR. LGTM

**Comment by Hangleton** [comment]
Thanks!

**Pull request merged**